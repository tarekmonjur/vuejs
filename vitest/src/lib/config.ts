import { readFileSync } from 'fs';

export const loadConfig = () => {
  return JSON.parse(readFileSync('./tsconfig.json', 'utf-8'));
};
