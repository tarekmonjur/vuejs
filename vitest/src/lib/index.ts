export const deepMerge = <T, T1 extends T>(a: T, b: T): Partial<T1> => {
  if (Array.isArray(a) && Array.isArray(b)) {
    return [...a, ...b] as T1;
  }

  const aType = Array.isArray(a) ? 'array' : 'object';
  const bType = Array.isArray(b) ? 'array' : 'object';
  if (aType !== bType) {
    throw new Error('Can not merge different types');
  }

  type mergeData = Partial<Record<any, any | T | T1>>;
  const merged = { ...a } as mergeData;

  for (const k of Object.keys(b as mergeData) as (keyof T)[]) {
    if (typeof a[k] === 'object' || Array.isArray(a[k])) {
      merged[k] = deepMerge(a[k], b[k]);
    } else {
      merged[k] = b[k];
    }
  }
  return merged;
};

export const gretting = (name: string) => {
  console.log('Hello, ' + name);
  console.log('Hi, ' + name);
};
