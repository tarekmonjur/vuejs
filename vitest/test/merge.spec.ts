import { describe, it, expect, test } from 'vitest';
import { deepMerge } from '../src/lib/index';

describe('shallow merge', () => {
  describe('merge two object data', () => {
    it('merge without overlap', () => {
      const results = deepMerge({ firstName: 'tarek' }, { lastName: 'monjur' });
      expect(results).toEqual({ firstName: 'tarek', lastName: 'monjur' });
    });

    it('merge with overlaps', () => {
      const results = deepMerge(
        { firstName: 'tarek', github: 'unknown' },
        { lastName: 'monjur', github: 'tarekmonjur' },
      );
      expect(results).toEqual({ firstName: 'tarek', lastName: 'monjur', github: 'tarekmonjur' });
    });
  });

  describe('merge two array data', () => {
    it('merge without overlap', () => {
      const results = deepMerge(['react', 'vue'], ['express', 'nest']);
      expect(results).toEqual(['react', 'vue', 'express', 'nest']);
    });
  });
});

describe('deep merge', () => {
  it('merge with overlap', () => {
    const results = deepMerge(
      { name: 'tarek', accounts: { github: 'tarekmonjur' } },
      { accounts: { twitter: 'tarek' } },
    );
    expect(results).toEqual({
      name: 'tarek',
      accounts: { github: 'tarekmonjur', twitter: 'tarek' },
    });
  });

  it('snapshorts merge with overlap', () => {
    const results = deepMerge(
      { name: 'tarek', accounts: { github: 'tarekmonjur' }, language: ['js'] },
      { accounts: { twitter: 'tarek' }, language: ['vue'] },
    );
    expect(results).toMatchSnapshot();
  });
});

test.fails('throw error on merge two different types', () => {
  const results = deepMerge(['foo', 'bar'], { foo: 'bar' });
  expect(results).toEqual({ 0: 'foo', 1: 'bar', foo: 'bar' });
});

test('Can not merge different types', () => {
  expect(() => deepMerge(['foo', 'bar'], { foo: 'bar' })).toThrowError(
    'Can not merge different types',
  );
});
