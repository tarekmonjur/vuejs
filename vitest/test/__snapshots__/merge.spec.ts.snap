// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`deep merge > snapshorts merge with overlap 1`] = `
{
  "accounts": {
    "github": "tarekmonjur",
    "twitter": "tarek",
  },
  "language": [
    "js",
    "vue",
  ],
  "name": "tarek",
}
`;
